### PR TITLE
Add internal solution for What's Up With Gravity (usaco-282)

### PR DESCRIPTION
**Pull request by ChuangSheep**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

File: solutions/usaco-282.mdx [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+```
+
+</CPPSection>
+</LanguageSection>
+
+
+<Info title="Alternative Solution">
+
+Since we are dealing with a shortest path problem, we can also use Dijkstra's
+algorithm to find the minimum "distance" (flips) needed to reach D. In this case,
```
Comment (SansPapyrus683) [inline solutions/usaco-282.mdx:140]:
```suggestion
algorithm to find the minimum flips needed to reach D. In this case,

File: solutions/usaco-282.mdx [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+there is an edge between two nodes if one of them can be reached by the other by
+going left, right or by flipping the gravity.
```
Comment (SansPapyrus683) [inline solutions/usaco-282.mdx:142]: wait how would dijkstra's be different from bfs

File: solutions/usaco-282.mdx [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+int main() {
+	freopen("gravity.in", "r", stdin);
+	freopen("gravity.out", "w", stdout);
+
+	int N, M;
+	cin >> N >> M;
+
+	pair<int, int> C, D;
```
Comment (SansPapyrus683) [inline solutions/usaco-282.mdx:87]: would prefer if you used variables like `start` and `end` instead

File: solutions/usaco-282.mdx [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+---
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_gravity.html)
+
+
+## Explanation
+
+Since this problem can be reduced to a shortest path problem with edge lengths of 0 or 1, we can use BFS.
+A transition edge has the weight 1 if the gravity is flipped or 0 if Captain Bovidian only moves left
+or right. We firstly add all reachable blocks from left and right to the queue.
```
Comment (SansPapyrus683) [inline solutions/usaco-282.mdx:15]: ok so in your impl your bfs queue has 4-tuples
maybe explain that here

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
yeah nice

File: solutions/usaco-282.mdx (hunk 2 above)
Comment (ChuangSheep) [reply solutions/usaco-282.mdx:142]: dijkstra uses a priority queue which will add log(n) to the time complexity.

File: solutions/usaco-282.mdx [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+ * @param flips Number of flips already performed to get to [r, c]
+ * @param direc Current direction of the gravity, either 1 or -1
+ */
+void search_horizontally(
+	int r, int c, int flips, int direc,
+	queue<array<int, 4>> &bfs,
+	vector<vector<int>> &flip_count,
+	const vector<vector<char>> &grid
+) {
+	if (r <= 0 || grid[r][c] == '-' || flip_count[r][c] >= 0) {
```
Comment (SansPapyrus683) [inline solutions/usaco-282.mdx:65]: maybe a comment to the effect of "if OOB, [whatever checking for a hypen does], or we've alr visited, don't go any further"

Comment (ChuangSheep) [reply solutions/usaco-282.mdx:66]: > whatever checking for a hypen does

It just checks out of bound

`// border is marked as '-'`

**Review by SansPapyrus683 — APPROVED** [review]

**Comment by ChuangSheep** [comment]
@envyaims pls ^ ~ ^

**Review by HarryWangATX — APPROVED** [review]

**Pull request merged**